enhancements/node-tuning/mixed-cpu-node-plugin.md

```diff
@@ -0,0 +1,352 @@
+that communicates with the container-runtime
+at various lifecycle stages of container e.g. CreateContainer, UpdateContainer, etc.,
+and modifies the container's cgroup to include a set of shared CPUs.
+
+Shared CPUs append to Kubelet [reservedSystemCpus](https://kubernetes.io/docs/tasks/administer-cluster/reserve-compute-resources/#explicitly-reserved-cpu-list).
+This means that when the feature enabled, Kubelet's `reservedSystemCpus` composed of PerformanceProfile's 
+`spec.cpu.reserved` + `spec.cpu.shared`.
+The reason we utilize Kubelet `reservedSystemCpus` is because of CPU manager, which is not aware of the CPUs
+lying in this pool, so it doesn't undo or changes the allocation logic performed by the plugin.
+Therefore, the plugin assigns the shared CPUs to containers without racing/conflict
+the CPU manager behavior. 
+More about this decision at the [Alternative](mixed-cpu-node-plugin.md#alternatives) section
+
+There is a plan for a feature called [shared-partition](https://github.com/openshift/enhancements/pull/1421) 
+that implements additional pool.
+When shared-partition lands, Kubelet's cpu pool layout should look roughly like:
+1. `reservedSystemCpus`
+2. `guaranteedCpus` <- new pool
+3. `sharedCpus` = All CPUs - `reservedSystemCpus` - `guaranteedCpus`
+
+When shared-partition feature enabled, 
+mixed-cpu-node-plugin allocates the shared cpus from the `sharedCpus` pool and not
+from the reservedSystemCpus anymore.
+
+The following section summarizes the shared-cpus source in each flow:
+* mixed-cpu enabled -> `cpu.shared` appends to Kubelet's `reservedSystemCPUs`
+* shared-partition enabled -> Kubelet's `sharedCpus` = All cpus - `reserved` - `isolated`
+* Both enabled -> sharedCpus = All cpus - `reserved` - `isolated`
+
+The node-plugin populates a special device named `openshift.io/shared-cpus` to provide way for pods to request
+for this special type of CPUs.
+There's no meaning to the value/number of `openshift.io/shared-cpus` devices that the pod requests.
+For example:
+```yaml
+requests:
+   openshift.io/shared-cpus: 1
+```
+Or
+```yaml
+requests:
+   openshift.io/shared-cpus: 2
+```
+Are the same in that context, the container has access to the same number of shared-cpus in both cases.
+In other words, the device request only uses as a hint to tell the plugin that shared-cpus required for the container and
+does not indicate any actual value.
+
+The reason for specifying a device, and not, for example, an annotation,
+is because when application pod requests for a device, the scheduler makes the pod pending till device up and running.
+This gives the node plugin room for setup and not being depended on pod admission order.
+So only when the node plugin finishes the registration process with NRI and device plugin, the scheduler
+admits the application pod.
+
+Once a container under the pod requests and gets access to shared cpus, new environment variable named
```

Comment:
what is the role of this env var?  why not a projected container?

what is the impact of this feature and other subsequent requests like in-place vertical scaling?

Reply:
Hi @derekwaynecarr thank you for taking the time and reviewing this EP. 
I'm still doing my due-diligence about some of the questions you raised but i'll try to address the one I can.

> what is the role of this env var? 

1. It tells the DPDK application user/developer, which of the cpus which the application's process has access to, are the shared ones.
The user can get the full list of accessible cpus using `taskset` or checking `proc/<pid>/self` and he also knows the shared cpus (by checking this env var). from here he can deduce the isolated cpus, and pin the application's threads according to his needs.
2. NRI needs to change the container cgroups's cpuset to include the shared cpus. it know on which containers it should apply this change, by looking for this env var: https://github.com/openshift-kni/mixed-cpu-node-plugin/blob/main/pkg/deviceplugin/deviceplugin.go#L64 

> why not a projected file?

Have you meant [projected volumes](https://kubernetes.io/docs/concepts/storage/projected-volumes/#:~:text=A%20projected%20volume%20maps%20several,downwardAPI)? 

> what is the impact of this feature and other subsequent requests like in-place vertical scaling?

Have you meant the feature introduced by this [KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/1287-in-place-update-pod-resources)?

Reply:
> what is the impact of this feature and other subsequent requests like in-place vertical scaling?

>> Have you meant the feature introduced by this [KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/1287-in-place-update-pod-resources)?

If we're talking about this feature, then the plugin should implement support in updating shared cpus in case it's being requested by an existing container.
This is possible and doesn't require any API changes at NRI level. 
We essentially should update the logic here:
https://github.com/openshift-kni/mixed-cpu-node-plugin/blob/main/pkg/nriplugin/nriplugin.go#L132